unify c2 and TH allocator

jerryzh168 opened the pull request:
Summary: Use the same function call structure in c2 allocator as in TH. i.e. moving allocation logic into a free function and separate functions into `alloc` and `alloc_internal`.

Differential Revision: D13202632

ezyang commented on aten/src/TH/THAllocator.cpp:
```diff
@@ -19,17 +19,7 @@
-};
-
-static THDefaultAllocator th_default_allocator;
+static DefaultCPUAllocator th_default_allocator;
```
I don't think you want this static variable anymore; it's a second copy of the allocator.

ezyang commented on c10/core/CPUAllocator.cpp:
```diff
@@ -0,0 +1,74 @@
+namespace c10 {
+
+void memset_junk(void* data, size_t num) {
+  // This garbage pattern is NaN when interpretted as floating point values,
```
nit: interpreted

ezyang commented on c10/core/CPUAllocator.cpp:
```diff
@@ -0,0 +1,74 @@
+#include <c10/DeviceType.h>
+
+C10_DEFINE_bool(
+    caffe2_report_cpu_memory_usage,
```
@dzhulgakov Do you know if anyone is actually using this? It seems like you would get waaay too much output, especially on PyTorch where we do lots of little allocations. Does this really warrant being compiled in by default; at the very most I would expect it to be a compile time flag?

dzhulgakov replied:
It's used in some benchamrks and toy examples.

jerryzh168 replied:
default is false so it should be fine?

ezyang replied:
Seems like a waste of code size 8) Oh well, not worth arguing about.

ezyang commented on c10/core/CPUAllocator.cpp:
```diff
@@ -0,0 +1,74 @@
+      static_cast<int64_t>(kJunkPattern) << 32 | kJunkPattern;
+  int32_t int64_count = num / sizeof(kJunkPattern64);
+  int32_t remaining_bytes = num % sizeof(kJunkPattern64);
+  int64_t* data_i64 = reinterpret_cast<int64_t*>(data);
```
FYI: This is a strict aliasing violation. You're supposed to use a `char*` pointer in this situation and use memcpy to copy the integers over.

ezyang replied:
cc @zheng-xq, original author of code

colesbury commented on c10/core/CPUAllocator.h:
```diff
@@ -0,0 +1,132 @@
+      return nullptr;
+    }
+
+    void* data;
```
Gottta love the placeholders in commit messages...

I didn't do any measurements to check if it was faster to not align small allocations. It was just the less risky approach at the time.

jerryzh168 replied:
Do we have benchmarks to measure the impact?

dzhulgakov replied:
cc @zou3519 for potential benchmarks to run

in general - it's safer to keep all tensor memory aligned (if it's larger than a cache line) - some kernels might rely on it

zou3519 replied:
https://github.com/pytorch/benchmark/tree/master/timing/cpp2 You could write a gbenchmark benchmark to test this

jerryzh168 commented on c10/core/CPUAllocator.h:
```diff
@@ -0,0 +1,132 @@
+namespace c10 {
+
+// Use 64-byte alignment should be enough for computation up to AVX512.
+constexpr size_t gCaffe2Alignment = 64;
```
Do we still use `CAFFE_ENFORCE` in c10?

dzhulgakov commented on c10/core/Allocator.cpp:
```diff
@@ -16,10 +16,6 @@ at::DataPtr InefficientStdFunctionContext::makeDataPtr(
-namespace caffe2 {
-
 C10_API at::Allocator* allocator_array[static_cast<int>(
     at::DeviceType::COMPILE_TIME_MAX_DEVICE_TYPES)];
```
we have already  casted constant: https://github.com/pytorch/pytorch/blob/master/c10/DeviceType.h#L33

dzhulgakov commented on aten/src/TH/THAllocator.cpp:
```diff
@@ -19,19 +19,8 @@
 
-struct THDefaultAllocator final : public at::Allocator {
-  at::DataPtr allocate(size_t size) const override {
-    auto* ptr = THAlloc(size);
```
shall we also route THAlloc itself to the new allocator?

dzhulgakov reviewed (commented):
There's also a legit failure in ASAN with
```
/var/lib/jenkins/workspace/aten/src/ATen/native/TensorFactories.cpp:117:16: runtime error: member call on address 0x7ff9ea481a80 which does not point to an object of type 'c10::Allocator'
Dec 01 03:26:37 0x7ff9ea481a80: note: object is of type 'c10::DefaultCPUAllocator'
```

I suspect it's because c10::Allocator class has all methods defined in .h file and it causes it to be weak symbols (including vtable pointer). And it confuses ASAN.

Can you move  destructor to cpp file please here? 
https://github.com/pytorch/pytorch/blob/master/c10/core/Allocator.h#L103

ezyang commented on aten/src/TH/THGeneral.cpp:
```diff
@@ -155,52 +159,12 @@ void THSetGCHandler( void (*torchGCFunction_)(void *data), void *data )
 {
-  void *ptr;
-
   if(size < 0)
```
Um, it probably doesn't matter, but why isn't this test moved into `alloc` proper?

jerryzh168 replied:
`alloc` is accepting size_t, which is unsigned integer.

ezyang commented on c10/core/Allocator.cpp:
```diff
@@ -2,6 +2,8 @@
 
 namespace c10 {
 
+Allocator::~Allocator() {}
```
Needs a comment

ezyang commented on c10/core/CPUAllocator.h:
```diff
@@ -0,0 +1,130 @@
+
+#include <c10/core/Allocator.h>
+#include "c10/util/Logging.h"
+#include "c10/util/numa.h"
```
nit: angled brackets, not quotes, please (yes it matters!)

ezyang reviewed (approved):
okey-dokey. Please check tests carefully.

ezyang commented on c10/core/CPUAllocator.cpp:
```diff
@@ -37,28 +37,66 @@ void memset_junk(void* data, size_t num) {
   }
 }
 
+void* alloc(size_t nbytes) {
```
nit: I don't think it's wise to take up such a "prime" name `c10::alloc` for aligned CPU allocation. It's not something most users really need from us, and we might need the name for something else at some later point in time.

ezyang commented on c10/core/CPUAllocator.cpp:
```diff
@@ -37,28 +37,66 @@ void memset_junk(void* data, size_t num) {
+      !FLAGS_caffe2_cpu_allocator_do_junk_fill)
+    << "Cannot request both zero-fill and junk-fill at the same time";
+  // Only fill data if we are running fbcode caffe2 code
+#ifdef C10_FBCODE_CAFFE2
```
If you are going to ifdef this, you should ifdef the check above, and you should also ifdef away the flags so that they're not available in OSS.

ezyang commented:
@jerryzh168 @dzhulgakov So, I know this was the conclusion y'all came to about how to best workaround the problem, but reading the patch as it stands, it is still the case that you are regressing allocations for Facebook-internal PyTorch users. I don't think people are going to be too happy about that, if they notice.

jerryzh168 commented:
@pytorchbot retest this please

jerryzh168 commented:
@pytorchbot retest this please

jerryzh168 committed:
unify c2 and TH allocator (#14517)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/14517

Merged caffe2 and TH CPU Allocators. Mostly using the code from caffe2 allocators.
Default caffe2 allocator will do memset, and default TH do not, right now we only turn on the memset when `FBCODE_CAFFE2` is enabled in order to avoid perf regression for pytorch users. We're planning to get rid of the memset in the future.

Baseline:
```
Running ./tensor_allocation
Run on (48 X 2501 MHz CPU s)
CPU Caches:
  L1 Data 32K (x24)
  L1 Instruction 32K (x24)
  L2 Unified 256K (x24)
  L3 Unified 30720K (x2)
-------------------------------------------------------------------------
Benchmark                                  Time           CPU Iterations
-------------------------------------------------------------------------
BM_MakeStorageImpl                       148 ns        148 ns    4676594
BM_StorageImplCtor                        54 ns         54 ns   12957810
BM_MallocStorageImpl                      62 ns         62 ns   11254745
BM_TensorImplCtor                         22 ns         22 ns   31939472
BM_MallocTensorImpl                      105 ns        105 ns    6505661
BM_Malloc_1                               43 ns         43 ns   16464905
BM_MakeTensorFromStorage                 126 ns        126 ns    5586116
BM_MakeVariableFromTensor                236 ns        236 ns    2995528
BM_ATenCPUTensorAllocationSmall1         319 ns        319 ns    2268884
BM_ATenCPUTensorAllocationSmall2         318 ns        318 ns    2163332
BM_ATenCPUTensorAllocationMedium1        403 ns        403 ns    1663228
BM_ATenCPUTensorAllocationMedium2        448 ns        448 ns    1595004
BM_ATenCPUTensorAllocationBig1           532 ns        532 ns    1352634
BM_ATenCPUTensorAllocationBig2          4486 ns       4486 ns     160978
```
Changed:
```
Running ./tensor_allocation
Run on (48 X 2501 MHz CPU s)
CPU Caches:
  L1 Data 32K (x24)
  L1 Instruction 32K (x24)
  L2 Unified 256K (x24)
  L3 Unified 30720K (x2)
-------------------------------------------------------------------------
Benchmark                                  Time           CPU Iterations
-------------------------------------------------------------------------
BM_MakeStorageImpl                       141 ns        141 ns    4803576
BM_StorageImplCtor                        55 ns         55 ns   13129391
BM_MallocStorageImpl                      64 ns         64 ns   11088143
BM_TensorImplCtor                         23 ns         23 ns   31616273
BM_MallocTensorImpl                      101 ns        101 ns    7017585
BM_Malloc_1                               39 ns         39 ns   18523954
BM_MakeTensorFromStorage                 118 ns        118 ns    5877919
BM_MakeVariableFromTensor                452 ns        452 ns    1565722
BM_ATenCPUTensorAllocationSmall1         384 ns        384 ns    1819763
BM_ATenCPUTensorAllocationSmall2         389 ns        389 ns    1857483
BM_ATenCPUTensorAllocationMedium1        425 ns        425 ns    1646284
BM_ATenCPUTensorAllocationMedium2        430 ns        430 ns    1561319
BM_ATenCPUTensorAllocationBig1           508 ns        508 ns    1309969
BM_ATenCPUTensorAllocationBig2          3799 ns       3799 ns     173674
```

lstm benchmark:
Before:
```
INFO:lstm_bench:Iter: 1 / 390. Entries Per Second: 0.7k.
INFO:lstm_bench:Iter: 21 / 390. Entries Per Second: 0.8k.
INFO:lstm_bench:Iter: 41 / 390. Entries Per Second: 0.8k.
INFO:lstm_bench:Iter: 61 / 390. Entries Per Second: 0.8k.
INFO:lstm_bench:Iter: 81 / 390. Entries Per Second: 0.8k.
INFO:lstm_bench:Iter: 101 / 390. Entries Per Second: 0.8k.
INFO:lstm_bench:Iter: 121 / 390. Entries Per Second: 0.8k.
INFO:lstm_bench:Iter: 141 / 390. Entries Per Second: 0.8k.
INFO:lstm_bench:Iter: 161 / 390. Entries Per Second: 0.8k.
INFO:lstm_bench:Iter: 181 / 390. Entries Per Second: 0.8k.
INFO:lstm_bench:Iter: 201 / 390. Entries Per Second: 0.8k.
INFO:lstm_bench:Iter: 221 / 390. Entries Per Second: 0.8k.
INFO:lstm_bench:Iter: 241 / 390. Entries Per Second: 0.7k.
INFO:lstm_bench:Iter: 261 / 390. Entries Per Second: 0.8k.
INFO:lstm_bench:Iter: 281 / 390. Entries Per Second: 0.8k.
INFO:lstm_bench:Iter: 301 / 390. Entries Per Second: 0.8k.
INFO:lstm_bench:Iter: 321 / 390. Entries Per Second: 0.8k.
INFO:lstm_bench:Iter: 341 / 390. Entries Per Second: 0.8k.
INFO:lstm_bench:Iter: 361 / 390. Entries Per Second: 0.8k.
INFO:lstm_bench:Iter: 381 / 390. Entries Per Second: 0.8k.
INFO:lstm_bench:Done. Total EPS excluding 1st iteration: 0.8k
```

After:
```
INFO:lstm_bench:Iter: 1 / 390. Entries Per Second: 0.8k.
INFO:lstm_bench:Iter: 21 / 390. Entries Per Second: 0.8k.
INFO:lstm_bench:Iter: 41 / 390. Entries Per Second: 0.8k.
INFO:lstm_bench:Iter: 61 / 390. Entries Per Second: 0.8k.
INFO:lstm_bench:Iter: 81 / 390. Entries Per Second: 0.8k.
INFO:lstm_bench:Iter: 101 / 390. Entries Per Second: 0.8k.
INFO:lstm_bench:Iter: 121 / 390. Entries Per Second: 0.8k.
INFO:lstm_bench:Iter: 141 / 390. Entries Per Second: 0.8k.
INFO:lstm_bench:Iter: 161 / 390. Entries Per Second: 0.8k.
INFO:lstm_bench:Iter: 181 / 390. Entries Per Second: 0.8k.
INFO:lstm_bench:Iter: 201 / 390. Entries Per Second: 0.8k.
INFO:lstm_bench:Iter: 221 / 390. Entries Per Second: 0.7k.
INFO:lstm_bench:Iter: 241 / 390. Entries Per Second: 0.7k.
INFO:lstm_bench:Iter: 261 / 390. Entries Per Second: 0.7k.
INFO:lstm_bench:Iter: 281 / 390. Entries Per Second: 0.7k.
INFO:lstm_bench:Iter: 301 / 390. Entries Per Second: 0.7k.
INFO:lstm_bench:Iter: 321 / 390. Entries Per Second: 0.7k.
INFO:lstm_bench:Iter: 341 / 390. Entries Per Second: 0.7k.
INFO:lstm_bench:Iter: 361 / 390. Entries Per Second: 0.7k.
INFO:lstm_bench:Iter: 381 / 390. Entries Per Second: 0.7k.
INFO:lstm_bench:Done. Total EPS excluding 1st iteration: 0.8k
```

Differential Revision: D13202632

fbshipit-source-id: 38c28855da66dd63273383da0af8a5062af3fd66
